### ISSUE TITLE
Quadratic time to first iteration in Stencil code

### ISSUE THREAD

**Issue by elliottslaughter** [StanfordLegion/legion#1316]
I'm seeing quadratic time to complete the first iteration of the Stencil code. I've run this in a couple of configurations now, on Crusher and Cori Haswell, `REALM_NETWORKS=mpi` and `REALM_NETWORKS=gasnetex`.

Here's an example from a set of runs I did on Crusher. This is `REALM_NETWORKS=gasnetex` with a `DEBUG=0` runtime, tracing, etc. The runs use 8 ranks per node but otherwise follow all our usual best practices and settings.

nodes | time to first iteration (seconds)
-- | --
1 | 0.0862
2 | 0.1616
4 | 0.4364
8 | 1.4192
16 | 5.1646
32 | 16.5056
64 | 66.02
128 | 302.2614

Here are profiles at 8 and 16 nodes for comparison:

  * 8 nodes (64 ranks): http://sapling.stanford.edu/~eslaught/crusher_startup_time_2022-08-30/legion_prof_8x8_r0
  * 16 nodes (128 ranks): http://sapling.stanford.edu/~eslaught/crusher_startup_time_2022-08-30/legion_prof_16x8_r0

What sticks out to me about these profiles is that `fill_` triggers a lot of `Send Index Space Request` on node 0.

Here's the region tree at 1 node (8 ranks): http://sapling.stanford.edu/~eslaught/crusher_startup_time_2022-08-30/region_graph.pdf

The region tree clearly shows that:

 1. Every region tree has at least one distjoint/complete partition.
 2. The specific partition used in the `fill_` launch is disjoint and complete. You can tell because the only other partition of the region is aliased, and would be rejected by both Regent and Legion.

For posterity, here is the corresponding application code. If you scroll back, you can see that each direction region (`x`/`y` * `p`/`m`) is partitioned twice and the disjoint partition goes into the `fill_` launch.

This is the `crusher-freeze-repro-2-gasnet` branch, which is based on Legion version 7d6c32e1fe36707a7e6b054002ed6835a232194a.

Has there been a regression in Legion? I'm really surprised that this application would see this behavior.

**Comment by lightsighter** [comment]
This looks to me like someone has asked Legion to compute either the disjointness or the completeness of a partition in the logical dependence analysis on the other nodes (hence the need to do something quadratic and all the index space communications). If I had to bet, then I would say that the application information about the disjointness or completeness of a partition is not making it to Legion.

**Comment by elliottslaughter** [comment]
The application did not specify completeness, so you're right that Legion is computing it. I'm checking what impact this has now.

**Comment by lightsighter** [comment]
Was that the case before? If we know the partitions are disjoint then it's usually not hard to check completeness.

**Comment by elliottslaughter** [comment]
At the time this code was originally written, Regent did not support `complete` and `incomplete` annotations. As a hack, I marked one incomplete partition as `aliased`: this is very important for Regent's SCR optimization in order to optimize the initialization of the SCR copy-in phase. Even though we are using DCR and not SCR, the `aliased` annotation has stuck around.

However, I just tried using correct annotations (all partitions correctly marked `disjoint` and with `complete` or `incomplete` as appropriate). The startup time has not changed.

I can get a new set of profiles.

**Comment by elliottslaughter** [comment]
For posterity, the change: https://gitlab.com/StanfordLegion/legion/-/commit/daa728f9321127b9ff340ba76dc146ac8350f88a

I ran with `-lg:partcheck` when I was debugging this and I'm pretty sure it's working because I initially made a few mistakes, and the part check caught them.

**Comment by lightsighter** [comment]
It would be better to make a reproducer on sapling so we can actually poke at it at a smaller scale. The quadratic behavior should still be there just less noticeable without a profiler.

**Comment by elliottslaughter** [comment]
Here's a profile on 8 nodes on Crusher:

http://sapling.stanford.edu/~eslaught/crusher_startup_time_2022-08-30/legion_prof_8x8_r0_complete

The pattern actually looks different now. Instead of being triggered by `fill_`, it is now being triggered by `stencil`. I'll have to double check the application, but it may be that the application does not write to the complete partition of the main grid initially.

**Comment by elliottslaughter** [comment]
No, I was wrong. Even for the main grid the application uses the complete partition exclusively; the incomplete partitions are only used for their bounds and are not used as region requirements anywhere in the code.

I'll get the reproducer on Sapling next.

**Comment by elliottslaughter** [comment]
If you go to `/scratch2/eslaught/legion_compute/language/stencil.run1_cpu_complete` there is a binary built against the `crusher-freeze-repro-2-gasnet` branch. You can run it with a command like:

```bash
salloc -N 1 -p gpu --exclusive
LD_LIBRARY_PATH=$PWD mpirun -n 2 ./stencil.idx -nx 2000 -ny 2000 -ntx 4 -nty 4 -tsteps 50 -tprune 30 -hl:sched 1024 -ll:cpu 4 -ll:util 1 -ll:bgwork 2 -ll:csize 15000  -ll:rsize 512
```

Feel free to change the `-ntx` and `-nty` depending on how many subregions you want to create.

If you want to rebuild, go to `/scratch2/eslaught/legion_compute/language` and run:

```bash
srun -N 1 -p gpu --exclusive --pty bash --login # to get on a compute node
CMAKE_PREFIX_PATH=/scratch2/eslaught/sw/llvm/llvm-11/install_g_nodes GASNET=$PWD/gasnet/release ./install.py --gasnet
./pldi22_scripts/build_stencil_cpu.sh stencil.run1_cpu_complete # choose a new directory for this so it doesn't conflict
```

**Comment by lightsighter** [comment]
This is a debug build?

**Comment by elliottslaughter** [comment]
No, do you want that?

**Comment by elliottslaughter** [comment]
If you do, you can either rebuilt (the `install.py` line, and add `--debug`), or let me know and I can do it.

**Comment by lightsighter** [comment]
If you can do it, then please do so and I'll look at it later today.

**Comment by elliottslaughter** [comment]
You can find it here: `/scratch2/eslaught/legion_compute/language/stencil.run1_cpu_debug`

**Comment by lightsighter** [comment]
I don't think this is a regression. If it is you'll need to point me at an earlier commit where it was performing better.

Here is what is happening: a partition is being made in a non-control-replicated task (e.g. not the top-level task). This partition has as many subspaces as there are processors. However, because it is made in a leaf task running on just one node, all of it's subregions are made on just one node (node 0). Now, when we get back to the top-level task and issue the `fill_` index task that is projecting over this newly created partition that was returned up the tree. Legion sees that it needs to issue a refinement operation to make the equivalence sets for this field of the region tree because it's never been used before. Legion attempts to shard these equivalence sets because the refinement is being done in a control replicated context and we want them to be distributed (all good). Now we hit the problem: all the index spaces for the partition have been made on node 0, so all the nodes start hammering node 0 asking for the specific subspaces that they need, and as you scale up, node 0 becomes a sequential bottleneck trying to distribute these subspaces out to all the nodes to make the equivalence sets. Moral of the story for the moment: make big partitions in control-replicated tasks, or control-replicate more levels of the task-tree (probably could use an important optimization here to make this more efficient if the parent task is also control replicated, but it should work alright).

**Comment by elliottslaughter** [comment]
I modified the application to `__demand(__inline)` the partition tasks. It required some Regent changes that I'll have to upstream. But in the `crusher-freeze-repro-2-gasnet` this appears to be fixed.

**Comment by lightsighter** [comment]
The optimization I had planned for this is not actually safe, so I'm not going to do it. Mentioning #481 as an example of a case where the profiler might be able to be smarter and decode the performance bug.
